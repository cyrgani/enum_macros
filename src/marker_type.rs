use alloc::vec;
use proc_macro2::{Span, TokenStream};
use quote::{format_ident, quote};
use syn::parse::{Parse, ParseStream};
use syn::punctuated::Punctuated;
use syn::{Fields, Generics, ItemEnum, Meta, Result, Token, Variant};

pub struct Arguments {
    attrs: Punctuated<Meta, Token![,]>,
}

impl Parse for Arguments {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(Self {
            attrs: input.parse_terminated(Meta::parse, Token![,])?,
        })
    }
}

pub fn marker_type(args: Arguments, item: ItemEnum) -> TokenStream {
    let marker = ItemEnum {
        variants: item
            .variants
            .iter()
            .map(|variant| Variant {
                fields: Fields::Unit,
                ..variant.clone()
            })
            .collect(),
        ident: format_ident!("{}Marker", item.ident, span = Span::call_site()),
        generics: Generics::default(),
        attrs: vec![],
        ..item.clone()
    };

    let attrs = args
        .attrs
        .into_iter()
        .map(|meta| {
            quote! {
                #[#meta]
            }
        })
        .collect::<TokenStream>();

    quote! {
        #item

        #[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
        #attrs
        #marker
    }
}
