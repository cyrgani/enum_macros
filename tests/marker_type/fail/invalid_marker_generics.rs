use enum_macros::marker_type;

#[marker_type]
enum Complex<T> {
    A(T),
    B { val: u8 },
}

struct F<T>(ComplexMarker<T>);

fn main() {}
