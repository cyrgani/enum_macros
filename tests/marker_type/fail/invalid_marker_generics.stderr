error[E0107]: enum takes 0 generic arguments but 1 generic argument was supplied
 --> tests/marker_type/fail/invalid_marker_generics.rs:9:13
  |
9 | struct F<T>(ComplexMarker<T>);
  |             ^^^^^^^^^^^^^--- help: remove these generics
  |             |
  |             expected 0 generic arguments
  |
note: enum defined here, with 0 generic parameters
 --> tests/marker_type/fail/invalid_marker_generics.rs:3:1
  |
3 | #[marker_type]
  | ^^^^^^^^^^^^^^
  = note: this error originates in the attribute macro `marker_type` (in Nightly builds, run with -Z macro-backtrace for more info)
